### PR TITLE
[extractor/litv] Fix playlist and video download

### LINKED ISSUE #5456: HTTP 405 on some of certain of site, eg: LiTV.
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a bug unrelated to a specific site
- [X] I've verified that I'm running yt-dlp version **2022.10.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/yt-dlp/yt-dlp/wiki/FAQ#video-url-contains-an-ampersand--and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Provide a description that is worded well enough to be understood

I have issue with HTTP 405 on certain video site with along by example: `https://testweb/myfolder`

If this happens on extractors, please modify like this so the videos downloads happily: `http://testweb/myfolder/`

These problem is documented [here](https://social.msdn.microsoft.com/Forums/zh-TW/5bd2162e-935c-4f8c-8656-f00ac09b9363/3553125945post26178652922998629983303404053767935492?forum=236) 

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
yt-dlp -vUcw --user-agent "Mozilla/5.0 (Windows NT 10.0.22623; Win64; arm64) HTTP/3.0 Chrome/107.0.5300.0 Edg/107.0.1379.1 yt-dlp/$VERSION aria2/$VERSION wget/$VERSION Transmission/3.00 IPTV Pro/$VERSION" --referer "%(referer)s" -N 4294967295 -f mergeall --video-multistreams --audio-multistreams --merge-output-format mp4 --embed-subs --sub-langs all --sub-format best --embed-thumbnail --embed-chapters --downloader aria2c "https://www.litv.tv/vod/drama/content.do?brc_id=root&id=VOD00219257&isUHEnabled=true&autoPlay=0" --yes-playlist
[debug] Command-line config: ['-vUcw', '--user-agent', 'Mozilla/5.0 (Windows NT 10.0.22623; Win64; arm64) HTTP/3.0 Chrome/107.0.5300.0 Edg/107.0.1379.1 yt-dlp/ aria2/ wget/ Transmission/3.00 IPTV Pro/', '--referer', '%(referer)s', '-N', '4294967295', '-f', 'mergeall', '--video-multistreams', '--audio-multistreams', '--merge-output-format', 'mp4', '--embed-subs', '--sub-langs', 'all', '--sub-format', 'best', '--embed-thumbnail', '--embed-chapters', '--downloader', 'aria2c', 'https://www.litv.tv/vod/drama/content.do?brc_id=root&id=VOD00219257&isUHEnabled=true&autoPlay=0', '--yes-playlist']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version 2022.10.04 [4e0511f] (pip) API
[debug] Python 3.10.8 (CPython 64bit) - Linux-4.9.186-22986754-aarch64-with-libc (libc)
[debug] Checking exe version: ffmpeg -bsfs
[debug] Checking exe version: ffprobe -bsfs
[debug] exe versions: ffmpeg 5.1.2 (setts), ffprobe 5.1.2
[debug] Optional libraries: Cryptodome-3.15.0, brotli-1.0.9, certifi-2022.06.15, mutagen-1.45.1, sqlite3-2.6.0, websockets-10.3
[debug] Proxy map: {}
[debug] Loaded 1690 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: 2022.10.04, Current version: 2022.10.04
yt-dlp is up to date (2022.10.04)
[debug] [LiTV] Extracting URL: https://www.litv.tv/vod/drama/content.do?brc_id=root&id=VOD00219257&isUHEnabled=true&autoPlay=0
[LiTV] VOD00219257: Downloading webpage
[LiTV] VOD00219257: Downloading JSON metadata
ERROR: [LiTV] VOD00219257: Unable to download JSON metadata: HTTP Error 405: Method Not Allowed (caused by <HTTPError 405: 'Method Not Allowed'>); please report this issue on
https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using
yt-dlp -U
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 672, in extract
ie_result = self._real_extract(url)
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/litv.py", line 98, in _real_extract
video_data = self._download_json(
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 1032, in download_content
res = getattr(self, download_handle.__name__)(url_or_request, video_id, **kwargs)
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 996, in download_handle
res = self._download_webpage_handle(
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 866, in _download_webpage_handle
urlh = self._request_webpage(url_or_request, video_id, note, errnote, fatal, data=data, headers=headers, query=query, expected_status=expected_status)
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 823, in _request_webpage
raise ExtractorError(errmsg, cause=err)
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/extractor/common.py", line 805, in _request_webpage
return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
File "/data/data/com.termux/files/usr/lib/python3.10/site-packages/yt_dlp/YoutubeDL.py", line 3682, in urlopen
return self._opener.open(req, timeout=self._socket_timeout)
File "/data/data/com.termux/files/usr/lib/python3.10/urllib/request.py", line 525, in open
response = meth(req, response)
File "/data/data/com.termux/files/usr/lib/python3.10/urllib/request.py", line 634, in http_response
response = self.parent.error(
File "/data/data/com.termux/files/usr/lib/python3.10/urllib/request.py", line 563, in error
return self._call_chain(*args)
File "/data/data/com.termux/files/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
result = func(*args)
File "/data/data/com.termux/files/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 405: Method Not Allowed
```

### PR DESCRIPTION
**Pull request by jiru**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This PR fixes the LiTV extractor, including playlist and video download. Note that all the videos on litv.tv are georestricted to Taiwan, while playlists are not.

Fixes #5456.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
🛠️🌟🎞️

Update `litv.py` extractor to handle new URL format, fix tests, and enhance metadata and playlist extraction. Use a different endpoint to get video URLs that works for more cases.

> _Oh we're the crew of the LiTV ship, and we sail the web so wide_
> _We fix the bugs and scrape the data, and we do it all with pride_
> _We changed the `URL` and the `test` and the `playlist` too_
> _And we heave away on the count of three, for the `video URL` is new_

### Walkthrough
*  Simplify and update playlist extraction for LiTV (`yt_dlp/extractor/litv.py`)
  - Use `content_id` instead of `id` as query parameter in `_URL_TEMPLATE` ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L16-R15))
  - Use `getProgramInfo` and `getSeriesTree` endpoints to get program and playlist data ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L71-R89))
  - Use `seriesId` as playlist id instead of `contentId` of first season ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L71-R89))
  - Use `getMainUrlNoAuth` endpoint instead of `getMainUrl` endpoint to avoid authentication ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L99-R100))
  - Check for meta refresh tag and raise error if content is not available ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L71-R89))
  - Simplify `_extract_playlist` method to take only two arguments and iterate over all seasons and episodes ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L40-R52))
  - Add `description`, `categories`, and `episode` fields to `info_dict` for second test case ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L33-R34))
  - Update `playlist_count` and `md5` values for second test case ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L24-R26))
  - Delete `traverse_obj` from import statement ([link](https://github.com/yt-dlp/yt-dlp/pull/7785/files?diff=unified&w=0#diff-2e2099b7b1d94e1b02c97214a6476e44a9d29d445e451acf456f1062fd98add5L8))

</details>

### CODE REVIEW

File: yt_dlp/extractor/litv.py [hunk 1]
```diff
@@ -5,59 +5,51 @@
...
-    }, {
-        'url': 'https://www.litv.tv/promo/miyuezhuan/?content_id=VOD00044841&',
-        'md5': '88322ea132f848d6e3e18b32a832b918',
-        'info_dict': {
-            'id': 'VOD00044841',
-            'ext': 'mp4',
-            'title': '芈月傳第1集　霸星芈月降世楚國',
-            'description': '楚威王二年，太史令唐昧夜觀星象，發現霸星即將現世。王后得知霸星的預言後，想盡辦法不讓孩子順利出生，幸得莒姬相護化解危機。沒想到眾人期待下出生的霸星卻是位公主，楚威王對此失望至極。楚王后命人將女嬰丟棄河中，居然奇蹟似的被少司命像攔下，楚威王認為此女非同凡響，為她取名芈月。',
-        },
-        'skip': 'Georestricted to Taiwan',
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:49]: why was this removed? it was already being skipped

File: yt_dlp/extractor/litv.py [hunk 2]
```diff
@@ -68,24 +60,33 @@ def _real_extract(self, url):
...
         # endpoint gives the same result as the data embedded in the webpage.
-        # If georestricted, there are no embedded data, so an extra request is
-        # necessary to get the error code
+        # If, for some reason, there are no embedded data, we do an extra request.
         if 'assetId' not in program_info:
             program_info = self._download_json(
                 'https://www.litv.tv/vod/ajax/getProgramInfo', video_id,
                 query={'contentId': video_id},
-                headers={'Accept': 'application/json'})
+                headers={'Content-Type': 'application/json',
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:77]: why send a `Content-Type` header if this is not a POST request?

File: yt_dlp/extractor/litv.py [hunk 3]
```diff
@@ -68,24 +60,33 @@ def _real_extract(self, url):
...
+                         'Accept': 'application/json'})
+
+        series_id = program_info['seriesId']
+        if self._yes_playlist(series_id, video_id, smuggled_data):
+            playlist_data = self._download_json(
+                'https://www.litv.tv/vod/ajax/getSeriesTree',
+                video_id,
+                query={'seriesId': series_id},
+                headers={'Content-Type': 'application/json',
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:86]: same question as above about the header

File: yt_dlp/extractor/litv.py [hunk 4]
```diff
@@ -68,24 +60,33 @@ def _real_extract(self, url):
 
         webpage = self._download_webpage(url, video_id)
 
+        if '<meta http-equiv="refresh" content="1;url=https://www.litv.tv/"' in webpage:
+            raise ExtractorError('No such content', expected=True)
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:64]: should we use regex to match slight variations in this pattern?

File: yt_dlp/extractor/litv.py [hunk 5]
```diff
@@ -5,59 +5,51 @@
...
-        },
-        'skip': 'Georestricted to Taiwan',
     }]
 
-    def _extract_playlist(self, season_list, video_id, program_info, prompt=True):
-        episode_title = program_info['title']
-        content_id = season_list['contentId']
+    def _extract_playlist(self, seasons_list, content_type):
+        episode_title = seasons_list['title']
+        content_id = seasons_list['contentId']
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:45]: Can we change the name of the `seasons_list` var since it is not actually a `list`? It makes the code hard to read. `playlist_data` would fine, since that is the name of the var being passed to it in `_real_extract()`

File: yt_dlp/extractor/litv.py [hunk 6]
```diff
@@ -5,59 +5,51 @@
...
 
         all_episodes = [
             self.url_result(smuggle_url(
-                self._URL_TEMPLATE % (program_info['contentType'], episode['contentId']),
+                self._URL_TEMPLATE % (content_type, episode['contentId']),
                 {'force_noplaylist': True}))  # To prevent infinite recursion
-            for episode in season_list['episode']]
+            for season in seasons_list['seasons']
+            for episode in season['episode']]
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:52]: We don't need 2 `for` loops:

```py
        all_episodes = [
            self.url_result(smuggle_url(
                self._URL_TEMPLATE % (content_type, episode['contentId']),
                {'force_noplaylist': True}))  # To prevent infinite recursion
            for episode in traverse_obj(playlist_data, ('seasons', ..., 'episode', lambda _, v: v['contentId']))]
```

**Review by bashonly — CHANGES_REQUESTED** [review]

**Comment by gamer191** [comment]
> Note that all the videos on litv.tv are georestricted to Taiwan, while playlists are not.

Huh? Can people bypass the geo-restrictions using playlists?

**Comment by bashonly** [comment]
> Huh? Can people bypass the geo-restrictions using playlists?

No, the videos are geo-blocked regardless. I think what jiru meant is that the playlist tests can pass for a dev who is not in region (because download tests for playlists only flat extract the playlist)

File: yt_dlp/extractor/litv.py (hunk 1 above)
Comment (jiru) [reply yt_dlp/extractor/litv.py:49]: I removed that particular test because the video it tries to download does not exist any more, so it will fail regardless of georestriction. I left tests that are (currently) passing if you use yt-dlp from Taiwan or using a proxy. In other words, you can still run the tests of this extractor after commenting out the `'skip': 'Georestricted to Taiwan',` lines.

Comment (bashonly) [reply yt_dlp/extractor/litv.py:49]: Ah I see. Can you add the test back, but change the `skip` value from "Georestricted to Tawain" to "No longer exists"

File: yt_dlp/extractor/litv.py (hunk 2 above)
Comment (jiru) [reply yt_dlp/extractor/litv.py:77]: Well, it looks like this header is not necessary for the extractor to work (although it does get sent by browsers to litv.tv on this request). I’ll remove it.

The 405 error did not come from this request. It came from the request I fixed in  fd49568d9d0074dd3e44aa6d4af919ddc7808d52, the endpoint simply got changed from `/vod/getMainUrl` to `/vod/ajax/getMainUrlNoAuth`.

File: yt_dlp/extractor/litv.py (hunk 4 above)
Comment (jiru) [reply yt_dlp/extractor/litv.py:64]: I bet this will never change, but yes, I can use a regex if you prefer.

**Comment by jiru** [comment]
@bashonly Thank you for reviewing my pull request! I think I have addressed all your comments, please let me know if there is anything else.

File: yt_dlp/extractor/litv.py [hunk 7]
```diff
@@ -46,20 +48,17 @@ class LiTVIE(InfoExtractor):
...
         all_episodes = [
             self.url_result(smuggle_url(
-                self._URL_TEMPLATE % (program_info['contentType'], episode['contentId']),
+                self._URL_TEMPLATE % (content_type, episode['contentId']),
                 {'force_noplaylist': True}))  # To prevent infinite recursion
-            for episode in season_list['episode']]
+            for episode in traverse_obj(playlist_data, ('seasons', ..., 'episode', lambda _, v: v['contentId']))]
 
-        return self.playlist_result(all_episodes, content_id, episode_title)
+        return self.playlist_result(all_episodes, playlist_data['contentId'], playlist_data['title'])
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:61]:
```suggestion
        return self.playlist_result(all_episodes, playlist_data['contentId'], playlist_data.get('title'))
```

File: yt_dlp/extractor/litv.py [hunk 8]
```diff
@@ -68,24 +67,33 @@ def _real_extract(self, url):
...
                 query={'contentId': video_id},
                 headers={'Accept': 'application/json'})
+
+        series_id = program_info['seriesId']
+        if self._yes_playlist(series_id, video_id, smuggled_data):
+            playlist_data = self._download_json(
+                'https://www.litv.tv/vod/ajax/getSeriesTree',
+                video_id,
+                query={'seriesId': series_id},
+                headers={'Accept': 'application/json'})
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:94]:
```suggestion
            playlist_data = self._download_json(
                'https://www.litv.tv/vod/ajax/getSeriesTree', video_id,
                query={'seriesId': series_id}, headers={'Accept': 'application/json'})
```

File: yt_dlp/extractor/litv.py [hunk 9]
```diff
@@ -68,24 +67,33 @@ def _real_extract(self, url):
 
         webpage = self._download_webpage(url, video_id)
 
+        if self._search_regex(
+                r'(?i)<meta\s[^>]*http-equiv="refresh"\s[^>]*content="[0-9]+;\s*url=https://www\.litv\.tv/"',
+                webpage, 'meta refresh redirect', default=False, group=0):
+            raise ExtractorError('No such content', expected=True)
```
Comment (bashonly) [inline yt_dlp/extractor/litv.py:73]:
```suggestion
            raise ExtractorError('No such content found', expected=True)
```

**Review by bashonly — APPROVED** [review]

File: yt_dlp/extractor/litv.py [hunk 10]
```diff
@@ -5,32 +5,34 @@
     ExtractorError,
     int_or_none,
     smuggle_url,
-    traverse_obj,
     unsmuggle_url,
+    traverse_obj,
```
Comment (Grub4K) [inline yt_dlp/extractor/litv.py:9]:
```suggestion
    traverse_obj,
    unsmuggle_url,
```

**Review by Grub4K — APPROVED** [review]
